itests/fe/service_accounts_test.py

```diff
@@ -126,3 +131,99 @@ def test_permission_grant_revoke(tmpdir, setup, browser):
+        grant_page.set_argument("bar")
+        grant_page.submit()
+
+        permission_rows = page.permission_rows
+        assert len(permission_rows) == 1
+        permission = permission_rows[0]
+        assert permission.permission == "other-permission"
+        assert permission.argument == "bar"
+
+
+def test_permission_grant_denied(tmpdir, setup, browser):
+    # type: (LocalPath, SetupTest, Chrome) -> None
+    with setup.transaction():
+        setup.add_user_to_group("gary@a.co", "some-group")
+        setup.grant_permission_to_group("some-permission", "foo", "some-group")
+        setup.create_service_account("service@svc.localhost", "some-group")
+
+    with frontend_server(tmpdir, "gary@a.co") as frontend_url:
+        browser.get(url(frontend_url, "/groups/some-group/service/service@svc.localhost/grant"))
+
+        page = ServiceAccountGrantPermissionPage(browser)
+        page.select_permission("some-permission (foo)")
+        page.set_argument("bar")
+        page.submit()
+
+        assert page.has_alert("The group some-group does not have that permission")
+
+
+def test_permission_grant_invalid_argument(tmpdir, setup, browser):
+    with setup.transaction():
+        setup.add_user_to_group("gary@a.co", "some-group")
+        setup.grant_permission_to_group("some-permission", "foo", "some-group")
+        setup.create_service_account("service@svc.localhost", "some-group")
+
+    with frontend_server(tmpdir, "gary@a.co") as frontend_url:
+        browser.get(url(frontend_url, "/groups/some-group/service/service@svc.localhost/grant"))
+
+        page = ServiceAccountGrantPermissionPage(browser)
+        page.select_permission("some-permission (foo)")
+        page.set_argument("@@@@")
+        page.submit()
+
+        print(page.root.page_source)
+        assert page.has_alert("argument")
+
+
+def test_permission_revoke_denied(tmpdir, setup, browser):
+    # type: (LocalPath, SetupTest, Chrome) -> None
+    with setup.transaction():
+        setup.create_service_account("service@svc.localhost", "some-group")
+        setup.grant_permission_to_service_account("some-permission", "*", "service@svc.localhost")
+        setup.create_user("gary@a.co")
+
+    with frontend_server(tmpdir, "gary@a.co") as frontend_url:
+        browser.get(url(frontend_url, "/groups/some-group/service/service@svc.localhost"))
+
+        page = ServiceAccountViewPage(browser)
+        assert page.owner == "some-group"
+        permission_rows = page.permission_rows
+        assert len(permission_rows) == 1
+        permission = permission_rows[0]
+        assert permission.permission == "some-permission"
+        assert permission.argument == "*"
+
+        # The button doesn't show for someone who can't manage the service account.
+        with pytest.raises(NoSuchElementException):
+            permission.click_revoke_button()
+
+    # Add the user to the group so that the revoke button will show up, and then revoke it before
```

Comment:
Very unimportant: I'd argue that testing this by hitting the URL directly is both a little easier and very marginally more correct, as in theory the current method could fail on some weird auto-update behavior with the endpoint remaining unauthed. Practically though doesn't matter.

Reply:
I completely agree and had started down that path and then discovered that Python Selenium doesn't allow you to do a test-initiated POST, so there's no way to hit the URL directly (since it's POST-only, since it mutates state) without having something to click on. I'll add a comment to make that clearer to the next person to look at this code.